types/express-serve-static-core/index.d.ts

```diff
@@ -40,27 +40,29 @@ export interface ParamsDictionary { [key: string]: string; }
 export type ParamsArray = string[];
 export type Params = ParamsDictionary | ParamsArray;
 
-export interface RequestHandler<P extends Params = ParamsDictionary, ResBody = any, ReqBody = any, ReqQuery = any> {
+export interface Query { [key: string]: string | string[] | Query | Query[]; }
```

Comment:
Maybe add a comment above here to say it's the return type of `qs.parse`, the default query parser.

Reply:
Hey @OliverJAsh, I just realized that the dictionary types don't take account when the `key` is not present. Should we update this to

```
export interface Query { [key: string]: string | string[] | Query | Query[] | undefined; }
```

Reply:
[In my opinion that's a bug in TypeScript](https://github.com/microsoft/TypeScript/issues/13778). I don't believe we should provide a workaround here, since it applies to all index signatures. If the user wants to workaround that, they can use a helper instead of indexing directly: https://github.com/microsoft/TypeScript/issues/13778#issuecomment-344637746

Reply:
Not including `undefined` though breaks this usage pattern for typing the request object of a specific route, which I had found quite useful:

```
interface MyRouteRequest extends Request {
  // Validation of the query object is in place so I'm basically asserting the type
  query: {
    requiredParam: string;
    optionalParam?: string;
  }
}
```

With `query: any` this didn't raise any compilation error, with `query: Query` (or `query: ParsedQs`, like it is now) this raises the following error, caused by the `?` of `optionalParam`:

```
Type 'undefined' is not assignable to type 'string | ParsedQs | string[] | ParsedQs[]'
```

I agree that the underlying problem is TypeScript's, though practically this change means that if before I could meaningfully and correctly type my query objects, now I can only go for a partially correct typing.

So I would suggest either adding `undefined` to the `ParsedQs` union type, or maybe better yet go back to using `any`. Rationale:

- as written above, it gives the developer the possibility to override the type
- `ParsedQs` is very broad as a type, so it doesn't give many more guarantees over `any`
- `ParsedQs` gives false guarantees about nullability (again, I agree it's a TypeScript problem, but at that point I would argue in favour of using `any`, which seems less deceptive)
- though `qs` is the default express query parser, express can use node's native query parser `querystring`, for which the `ParsedQs` typing would be incorrect

What do you think @puneetar , @OliverJAsh ? (If it sounds reasonable I'll open a PR for the change)

Reply:
I agree, that is a good reason to add `undefined` to the union. 👍 If someone wants to raise a PR for that I would be happy to approve it.

> better yet go back to using any

I strongly believe we should not go back to using `any` because it completely disables type checking—you have to opt-in by providing a generic, otherwise you may get runtime exceptions.

```ts
app.get('/', req => {
  // no compile error, but throws at runtime!
  req.query.i.do.not.exist
});
```

As soon as `any` is used (whether it's intentional or not), it may leak into other places and cause all sorts of bugs.

```ts
const fn = (foo: string) => foo.length;
app.get('/', req => {
  // no compile error, but throws at runtime if `foo` does not exist!
  fn(req.query.foo);
});
```

With the current model, type safety is built-in. If you don't care about type safety, you can opt-out.

> though `qs` is the default express query parser, express can use node's native query parser `querystring`, for which the `ParsedQs` typing would be incorrect

In that case you would use the generic to customise the query type to use `querystring.ParsedUrlQuery` instead. Alternatively you could write a function that asserts the type, then use that instead of reading `req.query` directly. That's actually what I'm doing:

```ts
/**
 * By default, Express uses `qs` to parse query strings:
 * https://expressjs.com/en/api.html#app.settings.table.
 *
 * However, we have switched this to use Node's built-in `querystring` parser instead. For this
 * reason, we need to assert the type.
 *
 * It is possible to change the type of `req.query` via a generic, however the generic is very
 * awkward to use since it requires many other unrelated generics to be provided at the same time.
 * */
export const getQueryForRequest = (req: express.Request) => req.query as ParsedUrlQuery;
```

Reply:
>Could you switch from extending Request to passing the custom query type as a generic? This PR added the query as a generic to support this.

Yeah, sorry, did not specify that which is the obvious answer. In theory it's possible, but the code I'm working with is a bit more convoluted, in that there are several layers of interface extension, each adding some custom properties to the request object. Something like:

```
interface ILoggedRequest extends Request {}
interface ITracedRequest extends IRequestWithLogger {}
interface IAuthenticatedRequest extends ITracedRequest {}
interface MyCustomRequest extends IAuthenticatedRequest {}
```

So I'd need to add generics to each step of the chain, some of which are defined in shared libraries, which makes it very complicated. Ofc it's not this package definition's problem, but if the proposed change makes sense for other use cases all the better.

I opened this PR https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44645 for it.